diff --git a/src/Makefile.am b/src/Makefile.am
index 5e5db4e..d5cea9c 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1,6 +1,8 @@
 ASSEMBLY = dbus-sharp
 TARGET = $(ASSEMBLY).dll
 XBUILD_Q = $(XBUILD) /nologo /verbosity:quiet
+SNKFILE = ../dbus-sharp.snk
+SN_R = sn -R
 
 # Use the gencssources script to generate this list
 CSFILES = Address.cs \
@@ -50,6 +52,7 @@ all: $(TARGET)
 
 $(TARGET): $(CSFILES) AssemblyInfo.cs
 	$(XBUILD_Q) dbus-sharp.csproj
+	$(SN_R) $(TARGET) $(SNKFILE)
 
 install-data-local: $(TARGET)
 	@if test -n '$(TARGET)'; then                       \
