diff --git a/mcs/class/Mono.Security/Mono.Security.X509/X509StoreManager.cs b/mcs/class/Mono.Security/Mono.Security.X509/X509StoreManager.cs
index 7f31713..15001fe 100644
--- a/mcs/class/Mono.Security/Mono.Security.X509/X509StoreManager.cs
+++ b/mcs/class/Mono.Security/Mono.Security.X509/X509StoreManager.cs
@@ -71,9 +71,7 @@ namespace Mono.Security.X509 {
 		internal static string LocalMachinePath {
 			get {
 				if (_localMachinePath == null) {
-					_localMachinePath = Path.Combine (
-						Environment.GetFolderPath (Environment.SpecialFolder.CommonApplicationData),
-						".mono");
+					_localMachinePath = Path.Combine ("/app",".mono");
 					_localMachinePath = Path.Combine (_localMachinePath, "certs");
 				}
 				return _localMachinePath;
@@ -95,9 +93,7 @@ namespace Mono.Security.X509 {
 		internal static string NewLocalMachinePath {
 			get {
 				if (_newLocalMachinePath == null) {
-					_newLocalMachinePath = Path.Combine (
-						Environment.GetFolderPath (Environment.SpecialFolder.CommonApplicationData),
-						".mono");
+					_newLocalMachinePath = Path.Combine ("/app",".mono");
 					_newLocalMachinePath = Path.Combine (_newLocalMachinePath, "new-certs");
 				}
 				return _newLocalMachinePath;
