diff --git a/src/AssemblyInfo.cs.in b/src/AssemblyInfo.cs.in
index 36137de..ace627f 100644
--- a/src/AssemblyInfo.cs.in
+++ b/src/AssemblyInfo.cs.in
@@ -4,7 +4,9 @@
 
 using System.Reflection;
 using System.Runtime.CompilerServices;
+using System.Runtime.Versioning;
 
+[assembly: TargetFramework(".NETFramework,Version=v4.0")]
 [assembly: AssemblyFileVersion("@VERSION@")]
 [assembly: AssemblyInformationalVersion("@VERSION@")]
 [assembly: AssemblyVersion("@API_VERSION@")]
