diff --git a/src/Build/project.json b/src/Build/project.json
index 3c04b31..95861d8 100644
--- a/src/Build/project.json
+++ b/src/Build/project.json
@@ -7,6 +7,7 @@
       "dependencies": {
         "Microsoft.VisualStudio.Setup.Configuration.Interop": "1.2.304-preview5",
         "System.Collections.Immutable": "1.3.1",
+        "System.Reflection.Metadata": "1.3.0",
         "System.Threading.Tasks.Dataflow": "4.5.24.0"
       }
     },
@@ -35,4 +36,4 @@
       }
     }
   }
-}
\ No newline at end of file
+}
diff --git a/src/MSBuild/project.json b/src/MSBuild/project.json
index 7d861f2..587a18c 100644
--- a/src/MSBuild/project.json
+++ b/src/MSBuild/project.json
@@ -5,7 +5,8 @@
   "frameworks": {
     "net46": {
       "dependencies": {
-        "Microsoft.VisualStudio.Setup.Configuration.Interop":  "1.2.304-preview5"
+        "Microsoft.VisualStudio.Setup.Configuration.Interop":  "1.2.304-preview5",
+        "System.Reflection.Metadata": "1.3.0"
       }
     },
     "netcoreapp1.0": {
diff --git a/src/Shared/AssemblyNameExtension.cs b/src/Shared/AssemblyNameExtension.cs
index 529335c..5fa15ba 100644
--- a/src/Shared/AssemblyNameExtension.cs
+++ b/src/Shared/AssemblyNameExtension.cs
@@ -12,7 +12,7 @@ using System.Collections.Generic;
 using System.Runtime.Serialization;
 #endif
 using System.IO;
-#if !FEATURE_ASSEMBLY_LOADFROM
+#if !FEATURE_ASSEMBLY_LOADFROM || MONO
 using System.Reflection.PortableExecutable;
 using System.Reflection.Metadata;
 #endif
@@ -130,7 +130,7 @@ namespace Microsoft.Build.Shared
         internal static AssemblyNameExtension GetAssemblyNameEx(string path)
         {
             AssemblyName assemblyName = null;
-#if FEATURE_ASSEMBLY_LOADFROM
+#if FEATURE_ASSEMBLY_LOADFROM && !MONO
             try
             {
                 assemblyName = AssemblyName.GetAssemblyName(path);
@@ -158,7 +158,16 @@ namespace Microsoft.Build.Shared
                 assemblyName = new AssemblyName();
                 assemblyName.Name = metadataReader.GetString(entry.Name);
                 assemblyName.Version = entry.Version;
-                assemblyName.CultureName = metadataReader.GetString(entry.Culture);
+                var cultureString = metadataReader.GetString(entry.Culture);
+                if (!NativeMethodsShared.IsMono)
+                {
+                    // set_CultureName throws NotImplementedException on Mono
+                    assemblyName.CultureName = cultureString;
+                }
+                else if (cultureString != null)
+                {
+                    assemblyName.CultureInfo = new CultureInfo(cultureString);
+                }
                 assemblyName.SetPublicKey(metadataReader.GetBlobBytes(entry.PublicKey));
                 assemblyName.Flags = (AssemblyNameFlags)(int)entry.Flags;
             }
diff --git a/src/Utilities/project.json b/src/Utilities/project.json
index 24fd438..bf52ecd 100644
--- a/src/Utilities/project.json
+++ b/src/Utilities/project.json
@@ -23,4 +23,4 @@
       }
     }
   }
-}
\ No newline at end of file
+}
diff --git a/targets/runtimeDependencies/project.json b/targets/runtimeDependencies/project.json
index 58d0b34..8127118 100644
--- a/targets/runtimeDependencies/project.json
+++ b/targets/runtimeDependencies/project.json
@@ -19,6 +19,7 @@
       "dependencies": {
         "xunit.runner.visualstudio": "2.1.0",
         "System.Collections.Immutable": "1.3.1",
+        "System.Reflection.Metadata": "1.3.0",
         "System.Threading.Tasks.Dataflow": "4.5.24.0",
         "Microsoft.DotNet.BuildTools.TestSuite": "1.0.0-prerelease-00508-01"
       }
