diff --git a/src/Shared/AssemblyNameExtension.cs b/src/Shared/AssemblyNameExtension.cs
index 529335c8..4b4c8277 100644
--- a/src/Shared/AssemblyNameExtension.cs
+++ b/src/Shared/AssemblyNameExtension.cs
@@ -859,6 +859,11 @@ namespace Microsoft.Build.Shared
             return asString;
         }
 
+        internal string ToString2()
+        {
+            return $"asAssemblyName: '{asAssemblyName}', asString: {asString}, isSimpleName: {isSimpleName}";
+        }
+
         /// <summary>
         /// Compare the fields of this with that if they are not null.
         /// </summary>
diff --git a/src/Tasks/AssemblyDependency/Resolver.cs b/src/Tasks/AssemblyDependency/Resolver.cs
index ef5b4c0e..6b12e687 100644
--- a/src/Tasks/AssemblyDependency/Resolver.cs
+++ b/src/Tasks/AssemblyDependency/Resolver.cs
@@ -132,6 +132,7 @@ namespace Microsoft.Build.Tasks
                 considered.SearchPath = searchPathElement;
             }
 
+            Console.WriteLine ($"MSB: ResolveAsFile fullPath: {fullPath}, assemblyName(Extension): {assemblyName?.ToString2()}");
             if (FileMatchesAssemblyName(assemblyName, isPrimaryProjectReference, wantSpecificVersion, allowMismatchBetweenFusionNameAndFileName, fullPath, considered))
             {
                 return true;
