diff --git a/main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/MonoTargetRuntimeFactory.cs b/main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/MonoTargetRuntimeFactory.cs
index 142c348d972..771a2601c5b 100644
--- a/main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/MonoTargetRuntimeFactory.cs
+++ b/main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/MonoTargetRuntimeFactory.cs
@@ -36,7 +36,7 @@ class MonoTargetRuntimeFactory: ITargetRuntimeFactory
 	{
 		static RuntimeCollection customRuntimes = new RuntimeCollection ();
 		static string configFile = UserProfile.Current.ConfigDir.Combine ("mono-runtimes.xml");
-		static string[] commonLinuxPrefixes = new string[] { "/usr", "/usr/local" };
+		static string[] commonLinuxPrefixes = new string[] { "/usr", "/usr/local", "/app" };
 		
 		static MonoTargetRuntimeFactory ()
 		{
