diff --git a/main/external/mono-addins/Mono.Addins/Mono.Addins.Database/FileDatabase.cs b/main/external/mono-addins/Mono.Addins/Mono.Addins.Database/FileDatabase.cs
index b933760..ce00ea6 100644
--- a/main/external/mono-addins/Mono.Addins/Mono.Addins.Database/FileDatabase.cs
+++ b/main/external/mono-addins/Mono.Addins/Mono.Addins.Database/FileDatabase.cs
@@ -451,7 +451,7 @@ public void WriteObject (string file, object obj, BinaryXmlTypeMap typeMap)
 		
 		string GetFileKey (string directory, string sharedFileName, string objectId)
 		{
-			int avlen = System.Math.Max (240 - directory.Length, 10);
+			int avlen = System.Math.Min (System.Math.Max (240 - directory.Length, 10), 130);
 			string name = sharedFileName + "_" + Util.GetStringHashCode (objectId).ToString ("x");
 			if (name.Length > avlen)
 				return name.Substring (name.Length - avlen);
