diff --git a/main/src/addins/CSharpBinding/CSharpBinding.csproj b/main/src/addins/CSharpBinding/CSharpBinding.csproj
index 59a1741..a20c96c 100644
--- a/main/src/addins/CSharpBinding/CSharpBinding.csproj
+++ b/main/src/addins/CSharpBinding/CSharpBinding.csproj
@@ -172,10 +172,8 @@
       <HintPath>..\..\..\build\bin\Microsoft.CodeAnalysis.Features.dll</HintPath>
       <Private>False</Private>
     </Reference>
-    <!-- NOTE:  On Mono this conflicts with the one in mscorlib -->
-    <Reference Include="System.ValueTuple" Condition="'$(OS)' != 'Unix'">
+    <Reference Include="System.ValueTuple">
       <HintPath>..\..\..\packages\System.ValueTuple.4.3.0\lib\netstandard1.0\System.ValueTuple.dll</HintPath>
-      <Aliases>SystemValueTuple</Aliases>
     </Reference>
   </ItemGroup>
   <ItemGroup>
