diff --git a/src/corehost/cli/fxr/CMakeLists.txt b/src/corehost/cli/fxr/CMakeLists.txt
index bb87c6a..2a59744 100644
--- a/src/corehost/cli/fxr/CMakeLists.txt
+++ b/src/corehost/cli/fxr/CMakeLists.txt
@@ -64,4 +64,4 @@ if (WIN32 AND CLI_CMAKE_PLATFORM_ARCH_ARM)
     target_link_libraries(hostfxr shell32.lib)
 endif()
 
-install_library_and_symbols (hostfxr)
+#install_library_and_symbols (hostfxr)
