Index: xamarin-pkg-core-setup/src/corehost/cli/fxr/CMakeLists.txt
===================================================================
--- xamarin-pkg-core-setup.orig/src/corehost/cli/fxr/CMakeLists.txt
+++ xamarin-pkg-core-setup/src/corehost/cli/fxr/CMakeLists.txt
@@ -9,7 +9,7 @@ if(WIN32)
     add_compile_options($<$<CONFIG:Release>:/MT>)
     add_compile_options($<$<CONFIG:Debug>:/MTd>)
 else()
-    add_compile_options(-fPIC)
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC" )
 endif()
 
 include(../setup.cmake)
Index: xamarin-pkg-core-setup/src/corehost/cli/setup.cmake
===================================================================
--- xamarin-pkg-core-setup.orig/src/corehost/cli/setup.cmake
+++ xamarin-pkg-core-setup/src/corehost/cli/setup.cmake
@@ -52,7 +52,7 @@ if(WIN32)
     set(CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO} /NODEFAULTLIB:libucrt.lib /DEFAULTLIB:ucrt.lib")
     set(CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} /NODEFAULTLIB:libucrt.lib /DEFAULTLIB:ucrt.lib")
 else()
-    add_compile_options(-Wno-unused-local-typedef)
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-local-typedef" )
 endif()
 
 # Older CMake doesn't support CMAKE_CXX_STANDARD and GCC/Clang need a switch to enable C++ 11
