--- a/daemon/Makefile.am
+++ b/daemon/Makefile.am
@@ -8,7 +8,6 @@
 	-D_LARGEFILE64_SOURCE                   \
 	-D_POSIX_PTHREAD_SEMANTICS		\
 	-D_REENTRANT				\
-	-DG_DISABLE_DEPRECATED			\
 	-DGNOME_VFS_PREFIX=\"$(prefix)\"	\
 	-DGNOME_VFS_DATADIR=\"$(datadir)\"	\
 	-DGNOME_VFS_LIBDIR=\"$(libdir)\"	\
--- a/libgnomevfs/Makefile.am
+++ b/libgnomevfs/Makefile.am
@@ -19,7 +19,6 @@
 	-D_LARGEFILE64_SOURCE                   \
 	-D_POSIX_PTHREAD_SEMANTICS		\
 	-D_REENTRANT				\
-	-DG_DISABLE_DEPRECATED			\
 	-DGNOME_VFS_PREFIX=\"$(prefix)\"	\
 	-DGNOME_VFS_BINDIR=\"$(bindir)\"	\
 	-DGNOME_VFS_DATADIR=\"$(datadir)\"	\
--- a/modules/Makefile.am
+++ b/modules/Makefile.am
@@ -18,7 +18,6 @@
 	-D_POSIX_PTHREAD_SEMANTICS		\
 	-D_REENTRANT				\
 	-DSSH_PROGRAM=\"$(SSH_PROGRAM)\"	\
-	-DG_DISABLE_DEPRECATED			\
 	-DGNOME_VFS_PREFIX=\"$(prefix)\"	\
 	-DGNOME_VFS_DATADIR=\"$(datadir)\"	\
 	-DGNOME_VFS_LIBDIR=\"$(libdir)\"	\
--- a/programs/Makefile.am
+++ b/programs/Makefile.am
@@ -4,8 +4,7 @@
 	-I$(top_srcdir)				\
 	-I$(top_builddir)			\
 	$(TEST_CFLAGS)				\
-	$(VFS_CFLAGS)				\
-	-DG_DISABLE_DEPRECATED
+	$(VFS_CFLAGS)
 
 libraries =						\
 	$(top_builddir)/libgnomevfs/libgnomevfs-2.la	\
--- a/test/Makefile.am
+++ b/test/Makefile.am
@@ -6,7 +6,6 @@
 	-I$(top_builddir)/libgnomevfs           \
 	$(TEST_CFLAGS)				\
 	$(VFS_CFLAGS)				\
-	-DG_DISABLE_DEPRECATED			\
 	-DMODULES_PATH=\"$(libdir)/vfs/modules\"
 
 if OS_WIN32
